### PR TITLE
Merge Tensor.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11970 Make ATen-core and caffe2 mutually recursive / merge template data<T>()&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D9967509/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11971 Merge TensorImpl.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D9995559/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11972 Merge UndefinedTensorImpl.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D9995633/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#11973 Merge Tensor.**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9996898/)

- Renamed at::Tensor from "struct" to "class", to line it up with
  Caffe2 (if it has private members, make it a class!)  Had to fix
  forward declarations.
- Fix forward declarations of caffe2::Tensor to forward declare
  at::Tensor and then declare using at::Tensor in caffe2
- Modify Tensor::data<T> to accept any type, not just ScalarType T.
- Rename tensor_impl_ field to impl_, to line up the impl name.
- Add qualifiers for some caffe2 only identifiers.
- Very critically, DELETE the now redundant CAFFE_KNOWN_TYPE for
  caffe2::Tensor. (Right now, this will silently cause very strange typeid
  mismatch errors.)
- Fix some newly added methods on Tensor which had the wrong type.
- Merge data() to correctly virtual dispatch, so that we can handle the PyTorch
  variable case. This is different from the old Caffe2 implementation, which
  was inlineable. To do this, I had to introduce some more helper methods,
  which do the obvious/logical things:
    - {Storage,StorageImpl}::IsTypeIdentifier(caffe2::TypeIdentifier)
    - {Storage,StorageImpl}::unsafe_data() (non-templated)
    - TypeIdentifier::name()

BC notes:
- Tensor::data<T>() now returns a non-const T*, not a const T* (as it did
  previously). This is mostly compatible, but in some cases we pass
  references to pointers, e.g., const T*&. This is NOT convertible.
  I generalize data to accept "const" types, and then modified call sites
  to pass const T to get a const pointer.
- New constructors for Tensor seem to cause some brace-enclosed
  initializer lists to become ambiguous. I fixed the one case where this
  happened; in the future, we may want to reduce the number of tensor
  constructors.

Differential Revision: [D9996898](https://our.internmc.facebook.com/intern/diff/D9996898/)

### CODE REVIEW

File: aten/src/ATen/core/TensorImpl.h [hunk 1]
```diff
@@ -110,6 +111,23 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
 
   explicit TensorImpl(at::Storage storage) : storage_(std::move(storage)), storage_offset_(0) {
     data_type_ = storage_ ? storage_.dtype() : caffe2::TypeMeta{};
+    if (storage_) {
```
Comment (gchanan) [inline aten/src/ATen/core/TensorImpl.h]: this looks fine to merge but IMO would should do things like:
1) take storage_offset for constructors taking Storage
2) Have a TensorTypeId for every device type we care about.
3) Take is_variable(?), TensorTypeId (?) for this as well.

The logic being this is an internal type, so the callers have to know what they want anyway, so put the complexity where it belongs (the weird constructors), not on the internal type.

File: aten/src/ATen/templates/Tensor.h [hunk 2]
```diff
@@ -259,35 +262,288 @@ struct CAFFE2_API Tensor {
...
 protected:
-  c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> tensor_impl_;
+  c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_;
+
+public:
+  explicit Tensor(Storage storage)
+      : impl_(c10::make_intrusive<TensorImpl, UndefinedTensorImpl>(std::move(storage))) {}
+
+  /**
+   * @brief Creates a tensor of the given dimension.
```
Comment (gchanan) [inline aten/src/ATen/templates/Tensor.h]: nit: dimensions.

File: aten/src/ATen/templates/Tensor.h [hunk 3]
```diff
@@ -259,35 +262,288 @@ struct CAFFE2_API Tensor {
...
+   *
+   * Note that the actual data allocation is not going to be carried out until
+   * the first time mutable_data() is called.
+   */
+  explicit Tensor(const std::vector<int64_t>& dims, DeviceType type)
```
Comment (gchanan) [inline aten/src/ATen/templates/Tensor.h]: why don't these take IntLists?

File: test/cpp/api/parallel.cpp [hunk 4]
```diff
@@ -55,7 +55,7 @@ TEST_F(ParallelTest, DifferentiableGather_MultiCUDA) {
   ASSERT_EQ(output.device(), torch::Device(torch::kCUDA, 1));
 
   auto chunks = output.chunk(2);
-  ASSERT_TRUE(chunks[0].to({torch::kCUDA, 0}).allclose(a));
+  ASSERT_TRUE(chunks[0].to(torch::Device{torch::kCUDA, 0}).allclose(a));
```
Comment (gchanan) [inline test/cpp/api/parallel.cpp]: does this not work anymore?

File: aten/src/ATen/templates/Tensor.h (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/templates/Tensor.h]: The "non-reason" is, that's just what Caffe2 had, I'm merging it :)

File: test/cpp/api/parallel.cpp (hunk 4 above)
Comment (ezyang) [reply test/cpp/api/parallel.cpp]: It doesn't. From the commit message:

New constructors for Tensor seem to cause some brace-enclosed
initializer lists to become ambiguous. I fixed the one case where this
happened; in the future, we may want to reduce the number of tensor
constructors.

**Comment by ezyang** [comment]
We will attempt this after we converge the APIs a little more.

**Pull request closed without merging**